Remove custom instance name and "Mark as finalized" checkbox

grzesiek2010 opened the pull request:
Closes #5416 

#### What has been done to verify that this works as intended?
I've tested the changes manually and added automated tests.

#### Why is this the best possible solution? Were any other approaches considered?
I've implemented the changes according to the acceptance criteria from the issue. I don't think there is a lot to discuss here.

~~There is just one thing that I would like to bring up and discuss... Upgrading the app and converting old settings into the new ones was considered in the issue and I've implemented it but what about scanning old QR codes with old settings `Default to finalized` and `Mark form as finalized`? Do you think we should also take it into account and convert those old settings into new ones in the same way we do when upgrading the app? Personally, I think it's not a must and we can skip it but just wanted to make sure you skipped that scenario deliberately and not by accident @seadowg.~~

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
This is one of the first issues from the [finalization flow](https://github.com/getodk/collect/issues?q=is%3Aissue+is%3Aopen+label%3A%22finalization+flow%22). Please read the acceptance criteria in the issue because it will be helpful to understand the changes and build tests cases. 
To me, the most tricky area to explore is scanning old QR codes (with old settings that I've removed) and upgrading from an app with the same old settings. Please tests all those cases carefully to make sure it works in the way it is described in the issue.

#### Do we need any specific form for testing your changes? If so, please attach one.
No.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/getodk/docs/issues/new) and include the link below.
No.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/docs/CODE-GUIDELINES.md#ui-components-style-guidelines)

grzesiek2010 committed:
Removed the 'Mark form as finalized' checkbox

grzesiek2010 committed:
Removed the instance name field

grzesiek2010 committed:
Removed the instance name info

grzesiek2010 committed:
Changed button text

grzesiek2010 committed:
Reworked the end screen layout using ConstraintLayout

grzesiek2010 committed:
Converted FormEndView to koltin

grzesiek2010 committed:
Added 'Save as draft' button

grzesiek2010 committed:
Removed the 'Default to finalized' option

grzesiek2010 committed:
Display an instance name in the End Screen

grzesiek2010 committed:
Fixed tests

grzesiek2010 committed:
Added an option to disable the 'Save as draft' button

grzesiek2010 committed:
Added an option to disable the 'Finalize' button

grzesiek2010 committed:
Converted FormEntryAccessPreferencesFragment to kotlin

grzesiek2010 committed:
Do not allow to disable both 'Save as draft' and 'Finalize'

grzesiek2010 committed:
Added tests to make sure both 'Save as draft' and 'Finalize' buttons can't be disabled at the same time

grzesiek2010 committed:
Implemented an extension function to get a non-null preference by key

grzesiek2010 committed:
Naming improvements

grzesiek2010 committed:
Converted FormFinalizingTest to kotlin

grzesiek2010 committed:
Added automated tests for 'Save as draft' and 'Finalize' options

seadowg reviewed (changes requested):
Probably worth dealing with conflicts before review here as it involves a key class.

grzesiek2010 committed:
Naming improvements

seadowg commented:
> There is just one thing that I would like to bring up and discuss... Upgrading the app and converting old settings into the new ones was considered in the issue and I've implemented it but what about scanning old QR codes with old settings Default to finalized and Mark form as finalized? Do you think we should also take it into account and convert those old settings into new ones in the same way we do when upgrading the app? Personally, I think it's not a must and we can skip it but just wanted to make sure you skipped that scenario deliberately and not by accident @seadowg.

No, that was definitely left out by mistake. We'll need old QR codes to be migrated as well - they already are for other settings migrations.

grzesiek2010 committed:
Fixed conflicts

grzesiek2010 committed:
Migrate settings

grzesiek2010 committed:
Naming improvements

grzesiek2010 commented on settings/src/main/java/org/odk/collect/settings/ODKAppSettingsMigrator.java:
```diff
@@ -124,11 +126,13 @@ private List<Migration> getUnprotectedMigrations() {
-                moveKey("knownUrlList").toPreferences(metaPrefs)
+                moveKey("knownUrlList").toPreferences(metaSettings),
+
+                moveKey("default_completed").toPreferences(protectedSettings)
```
I moved this key to protected settings first so that it's easy to perform the migrations (all keys are in the same settings). Otherwise, I would need to implement something like a cross-settings migration which would be more complex.

grzesiek2010 commented:
> No, that was definitely left out by mistake. We'll need old QR codes to be migrated as well - they already are for other settings migrations.

Ok this is now fixed and I've updated the acceptance criteria in the issue.

seadowg commented on collect_app/src/main/java/org/odk/collect/android/injection/config/AppDependencyModule.java:
```diff
@@ -327,7 +327,9 @@ public ODKAppSettingsImporter providesODKAppSettingsImporter(Context context, Pr
                 projectsRepository,
                 settingsProvider,
                 Defaults.getUnprotected(),
+                Defaults.getRemovedUnprotected(),
```
Do we need this so that we don't end up in a weird state if the settings don't exist? Like when upgrading from a new (after this) version for instance. I can't seem to make any test fail by passing in `new HashMap<>()` instead.

grzesiek2010 replied:
Oh, it's redundant you are right. I've implemented it this way first and then improved [ODKAppSettingsMigrator](https://github.com/getodk/collect/pull/5556/files#diff-e0f2d4b940734742dc5079c53e36d7ee20029640cc8120debd85ca03cd0a77d6R164) by adding:

```
                        .withValues(false, null)
                        .toPairs(
                                ProtectedProjectKeys.KEY_SAVE_AS_DRAFT, false,
                                ProtectedProjectKeys.KEY_FINALIZE, true
                        ),
```

and didn't remove that. It's fixed now.

grzesiek2010 committed:
Removed unused import left after fixing conflicts

grzesiek2010 committed:
Get rid of loading removed defaults

seadowg commented on androidshared/src/main/java/org/odk/collect/androidshared/utils/PreferenceFragmentCompatUtils.kt:
```diff
@@ -0,0 +1,8 @@
+import androidx.preference.Preference
+import androidx.preference.PreferenceFragmentCompat
+
+fun <T : Preference> PreferenceFragmentCompat.getPreference(key: String): T {
```
Maybe this should live in `testshared` seeing as it's only used in tests? Or, do you think it'll be useful in non-test code as well?

grzesiek2010 replied:
This could be used in non-test code but I didn't want to refactor the existing code as a part of this pr.

seadowg commented on collect_app/src/androidTest/java/org/odk/collect/android/support/pages/FormEndPage.java:
```diff
@@ -27,46 +25,41 @@ public FormEndPage assertOnPage() {
-    public <D extends Page<D>> D clickSaveAndExit(D destination) {
-        onView(withId(R.id.save_exit_button)).perform(click());
+    public <D extends Page<D>> D clickSaveAsDraft(D destination) {
+        onView(withId(R.id.save_as_draft)).perform(click());
```
The `onView` matchers for "Save as draft" and "Finalize" could probably use `withText` instead of `withId`.

grzesiek2010 replied:
Done.

seadowg commented on collect_app/src/main/res/layout/form_entry_end.xml:
```diff
@@ -17,116 +17,53 @@ the specific language governing permissions and limitations under the License.
-            android:text="@string/save_form_as"
+        <com.google.android.material.button.MaterialButton
+            android:id="@+id/save_as_draft"
+            style="@style/Widget.Collect.Button.FullWidth"
```
Let's just use Material 3 styles for these buttons for the moment, and then we can tweak them before release once the whole new experience has been built out. I think the "Save as Draft" button can use `Widget.Material3.Button.OutlinedButton` (we'll need to define a theme attribute for whatever the outline uses) and "Finalize" can just use `Widget.Material3.Button`.

grzesiek2010 replied:
Done.

seadowg commented on collect_app/src/androidTest/java/org/odk/collect/android/feature/formentry/FormFinalizingTest.kt:
```diff
@@ -0,0 +1,98 @@
+            .copyForm(FORM)
+            .startBlankForm("One Question")
+            .swipeToEndScreen()
+            .clickFinalize(FormEndPage("One Question"))
```
You should be able to use `assertDisabled` here and in the test above.

grzesiek2010 replied:
Done.

seadowg commented on collect_app/src/main/java/org/odk/collect/android/activities/FormFillingActivity.java:
```diff
@@ -998,7 +993,7 @@ public boolean onOptionsItemSelected(MenuItem item) {
             case R.id.menu_save:
                 // don't exit
-                saveForm(false, InstancesDaoHelper.isInstanceComplete(false, settingsProvider.getUnprotectedSettings().getBoolean(KEY_COMPLETED_DEFAULT), getFormController()), null, true);
+                saveForm(false, InstancesDaoHelper.isInstanceComplete(false, false, getFormController()), null, true);
```
It looks like we can remove most of the parameters from `isInstanceComplete` now, as we seem to always be passing the same things in. It might be that that will let us move the method over to `FormController` easily as well.

grzesiek2010 replied:
Done.

seadowg commented on settings/src/main/resources/client-settings.schema.json:
```diff
@@ -315,10 +312,13 @@
           "type": "boolean"
         },
-        "save_as": {
+        "mark_as_finalized": {
```
Should `mark_as_finalized` be removed?

grzesiek2010 replied:
No, if we remove it scanning old QR codes won't work as expected because this option will not be recognized and will be skipped.

grzesiek2010 replied:
Probably we should have tests that use the real scheme to cover such cases right?

seadowg replied:
Ah, so we need the schema to recognize the old fields so that we can migrate them?

grzesiek2010 replied:
Correct. I've added a new test for this: https://github.com/getodk/collect/pull/5556/commits/0772f89b92e6debe5e913da5c0ec4ba43c505dac

seadowg commented:
@getodk/testers there are a few tweaks needed here, but I think this is ready for testing as there's nothing big.

grzesiek2010 committed:
Removed unused methods from FormEndPage

grzesiek2010 committed:
Use withText instead of withId in the onView matchers

grzesiek2010 committed:
Cleaned the isInstanceComplete method

grzesiek2010 committed:
Improved checking if button is disabled

grzesiek2010 committed:
Improved styling

grzesiek2010 committed:
Removed unused import and style

seadowg commented on collect_app/src/main/res/layout/form_entry_end.xml:
```diff
@@ -17,116 +17,53 @@ the specific language governing permissions and limitations under the License.
+        <com.google.android.material.button.MaterialButton
+            android:id="@+id/save_as_draft"
+            style="Widget.Material3.Button.OutlinedButton"
+            android:backgroundTint="@color/surfaceButtonColor"
```
You shouldn't need the `backgroundTint` or `textColor` now that this is using the `OutlinedButton` style.

grzesiek2010 replied:
Done.

seadowg commented on collect_app/src/main/res/layout/form_entry_end.xml:
```diff
@@ -17,116 +17,53 @@ the specific language governing permissions and limitations under the License.
-            android:text="@string/save_form_as"
+        <com.google.android.material.button.MaterialButton
+            android:id="@+id/save_as_draft"
+            style="Widget.Material3.Button.OutlinedButton"
```
This needs to be `@style/Widget.Material3.Button.OutlinedButton` for it to work. Same for `finalize`.

It looks like we also need to define the Material 3 "Outline" color for this to work. For the moment, I reckon we just go with a medium emphasis "On Surface" color:

```xml
<item name="colorOutline">@color/color_on_surface_medium_emphasis</item>
```

grzesiek2010 replied:
Done.

grzesiek2010 committed:
Styling improvements

grzesiek2010 committed:
Added new tests

grzesiek2010 committed:
Fixed button style

seadowg commented:
As discussed (with @lognaturel and @grzesiek2010), merging this pre QA and any issues found can be filed as bugs.

dbemke commented:
The layout of the text "You are at the end..." has changed. In the previous version it looks like it was aligned to the center. In the new design it looks as if it was aligned to the left. I think the previous layout looks a bit more organized.

![youareattheend](https://user-images.githubusercontent.com/56479916/236778948-58d6c0e9-440b-4c25-b266-230d08b3b197.png)

grzesiek2010 commented:
> The layout of the text "You are at the end..." has changed. In the previous version it looks like it was aligned to the center. In the new design it looks as if it was aligned to the left. I think the previous layout looks a bit more organized.

This is an intended change.

srujner commented:
**Tested with success!**

**Verified on device with Android 12 & 13**

**Verified cases:**

- All of the Acceptance Criteria from #5416 has been checked 
- Regression checks and Exploratory testing on Project settings and user settings; 
- Adding project manually or through QR code; 
- Switching between projects, adding duplicated projects; 
- Update settings from older version; 
- Invalid/Inverted QR code,

dbemke commented:
**Tested with success!**

Verified on device with Android 10
